benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py

```diff
@@ -0,0 +1,143 @@
+import torch
+import torch.distributed as c10d
+import torch.nn as nn
+from servers.AverageParameterServer import AverageParameterServer
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+from .DdpTrainerBase import DdpTrainerBase
+from .RpcTrainerBase import RpcTrainerBase
+
+
+class DdpSparseRpcTrainer(DdpTrainerBase, RpcTrainerBase):
+
+    PS_MAP = {
+        "AverageParameterServer": AverageParameterServer
+    }
+
+    class HookState:
+
+        def __init__(self, cref, process_group):
+            self.cref = cref
+            self.process_group = process_group
+            self.process_group_size = process_group.size()
+            self.param_loc = 0
+            self.batch_number = -1
+
+        def get_key(self):
+            return f"{self.batch_number},{self.param_loc}"
+
+        def next_batch_state(self):
+            self.param_loc = 0
+            self.batch_number += 1
+
+    def __init__(self, rank, trainer_count, ps_rref, ps_name, backend, use_cuda_rpc, epochs):
+        super().__init__(rank)
+        self.rank = rank
+        self.trainer_count = trainer_count
+        self.ps_rref = ps_rref
+        self.ps = self.PS_MAP[ps_name]
+        self.backend = backend
+        self.use_cuda_rpc = use_cuda_rpc
+        self.epochs = epochs
+
+    @staticmethod
+    def get_tensor_fut(bucket):
+        fut = torch.futures.Future()
+        fut.set_result([bucket.get_tensor()])
+        return fut
+
+    @staticmethod
+    def process_bucket(state, bucket):
+        cref = state.cref
+        tensor = bucket.get_tensor()
+        tensors_count = len(cref.bucket_to_parameters(bucket))
+        sparse = tensor.is_sparse
+        if not cref.use_cuda_rpc:
+            tensor = tensor.cpu()
+        if sparse:
+            tensor = cref.sparse_tensor_to_rpc_format(tensor)
+        if state.batch_number > 0:
+            ps = cref.ps
+            ps_args = [
+                cref.ps_rref,
+                state.batch_number,
+                state.param_loc,
+                tensor
+            ]
+            fut = cref.send_async_request(
+                state.get_key(),
+                cref.ps_rref,
+                ps.average_gradient,
+                *ps_args
+            )
+        else:
+            fut = cref.get_tensor_fut(bucket)
+        state.param_loc += tensors_count
+        return fut
+
+    @staticmethod
+    def hook(state, bucket):
+        cref = state.cref
+        tensor = bucket.get_tensor()
+        tensors_count = len(cref.bucket_to_parameters(bucket))
+        if tensor.is_sparse:
+            return cref.process_bucket(state, bucket)
```

Comment:
issue created to track work for fixing RuntimeError https://github.com/pytorch/pytorch/issues/58255

Reply:
the issue is closed